### ISSUE TITLE
InterfaceOptionsFrame_OpenToCategory requires two calls to open addon settings

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#89]
This is an ancient bug that's never been fixed and affects all clients. Given the example case below, the first call to `InterfaceOptionsFrame_OpenToCategory` will open the Interface options window but will incorrectly not swap to the requested category, instead just opening to the default "Controls" one in the "Game" tab.

```lua
local Panel = CreateFrame("Frame", "MyAddonPanel", InterfaceOptionsFramePanelContainer);
Panel.name = "Test";
InterfaceOptions_AddCategory(Panel);
InterfaceOptionsFrame_OpenToCategory("Test");
```

If an additional call to the function is provided, things work fine.

```lua
local Panel = CreateFrame("Frame", "MyAddonPanel", InterfaceOptionsFramePanelContainer);
Panel.name = "Test";
InterfaceOptions_AddCategory(Panel);
InterfaceOptionsFrame_OpenToCategory("Test");
InterfaceOptionsFrame_OpenToCategory("Test");  -- Added, this works around the issue.
```

The fundamental issue I believe is related to [this loop in InterfaceOptionsFrame_OpenToCategory](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/FrameXML/InterfaceOptionsFrame.lua#L467-L474):

```lua
		local buttons = InterfaceOptionsFrameAddOns.buttons;
		for i, button in SecureNext, buttons do
			if ( button.element == elementToDisplay ) then
				InterfaceOptionsListButton_OnClick(button);
			elseif ( elementToDisplay.parent and button.element and (button.element.name == elementToDisplay.parent and button.element.collapsed) ) then
				OptionsListButtonToggle_OnClick(button.toggle);
			end
		end
```

When the initial call to OpenToCategory occurs none of the buttons in the `InterfaceOptionsFrameAddOns.buttons` table have their `element` field populated, as a result we never enter into the [InterfaceOptionsList_DisplayPanel function](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/FrameXML/InterfaceOptionsFrame.lua#L18-L30) which is responsible for setting the `displayedPanel` field on the container, and when the interface options frame is shown at the end of the OpenToCategory function the [OnShow handler](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/FrameXML/InterfaceOptionsFrame.lua#L385-L395) sees that the `displayedPanel` field is unset and proceeds to take us to the wrong place.

If a call to [InterfaceAddOnsList_Update](https://github.com/Gethe/wow-ui-source/blob/live/FrameXML/InterfaceOptionsFrame.lua#L138-L193) is inserted prior to opening the panel, the  `element` field will be properly set and the first call to OpenToCategory will actually work as intended.

```lua
local Panel = CreateFrame("Frame", "MyAddonPanel", InterfaceOptionsFramePanelContainer);
Panel.name = "Test";
InterfaceOptions_AddCategory(Panel);
InterfaceAddOnsList_Update();
InterfaceOptionsFrame_OpenToCategory("Test");
```

**Comment by Anonomit** [comment]
There exists another case where a user has so many addons that they would need to scroll down in the list to see the settings that they want. When this happens, no amount of calls are enough to open the addon options correctly. The scroll bar needs to be manually adjusted so that the options category is visible.

I don't know if it still works, but here is an old fix. Line at the bottom could be uncommented to fix the issue for all addons.
```lua
-- Fix InterfaceOptionsFrame_OpenToCategory not actually opening the category (and not even scrolling to it)
-- Taken from addon BlizzBugsSuck (https://www.wowinterface.com/downloads/info17002-BlizzBugsSuck.html) and edited to not be global
do
	local function get_panel_name(panel)
		local tp = type(panel)
		local cat = INTERFACEOPTIONS_ADDONCATEGORIES
		if tp == "string" then
			for i = 1, #cat do
				local p = cat[i]
				if p.name == panel then
					if p.parent then
						return get_panel_name(p.parent)
					else
						return panel
					end
				end
			end
		elseif tp == "table" then
			for i = 1, #cat do
				local p = cat[i]
				if p == panel then
					if p.parent then
						return get_panel_name(p.parent)
					else
						return panel.name
					end
				end
			end
		end
	end

	local function InterfaceOptionsFrame_OpenToCategory_Fix(panel)
		if doNotRun or InCombatLockdown() then return end
		local panelName = get_panel_name(panel)
		if not panelName then return end -- if its not part of our list return early
		local noncollapsedHeaders = {}
		local shownpanels = 0
		local mypanel
		local t = {}
		local cat = INTERFACEOPTIONS_ADDONCATEGORIES
		for i = 1, #cat do
			local panel = cat[i]
			if not panel.parent or noncollapsedHeaders[panel.parent] then
				if panel.name == panelName then
					panel.collapsed = true
					t.element = panel
					InterfaceOptionsListButton_ToggleSubCategories(t)
					noncollapsedHeaders[panel.name] = true
					mypanel = shownpanels + 1
				end
				if not panel.collapsed then
					noncollapsedHeaders[panel.name] = true
				end
				shownpanels = shownpanels + 1
			end
		end
		local Smin, Smax = InterfaceOptionsFrameAddOnsListScrollBar:GetMinMaxValues()
		if shownpanels > 15 and Smin < Smax then
			local val = (Smax/(shownpanels-15))*(mypanel-2)
			InterfaceOptionsFrameAddOnsListScrollBar:SetValue(val)
		end
		doNotRun = true
		InterfaceOptionsFrame_OpenToCategory(panel)
		doNotRun = false
	end

	--hooksecurefunc("InterfaceOptionsFrame_OpenToCategory", InterfaceOptionsFrame_OpenToCategory_Fix)
end
```

**Comment by Meorawr** [comment]
This'll solve itself in 10.0 anyway.

**Comment by Meorawr** [comment]
Closing as this is basically fixed in 10.0 - whether or not Classic ever gets the new settings infrastructure is an unknown, but it certainly won't get a fix for this specific issue.
